File: c#/src/Sodium/Sodium/ImmutableCompositeListener.cs
using System.Collections.Generic;

namespace Sodium
{
    public class ImmutableCompositeListener : IListener, IWeakListener
    {
        private readonly IReadOnlyList<IListener> listeners;

        public ImmutableCompositeListener(IReadOnlyList<IListener> listeners)
        {
            this.listeners = listeners;
        }

        public void Unlisten()
        {
            foreach (IListener l in this.listeners)
            {
                l?.Unlisten();
            }
        }

        public IWeakListener GetWeakListener()
        {
            return this;
        }
    }
}

File: swift/src/Sodium/Sodium/Sodium/Sodium/ImmutableCompositeListener.swift
public class ImmutableCompositeListener : Listener
{
    private let listeners: [Listener]

    public init(listeners: [Listener]?)
    {
        self.listeners = listeners ?? []
        super.init(unlisten: {})
    }

    deinit
    {
        self.Unlisten()
    }

    public override func Unlisten() {
        for l in self.listeners {
            l.Unlisten()
        }
    }
}
